pkg/api/v1/control-plane/cache/prioritylist.go

```diff
@@ -0,0 +1,165 @@
+package cache
+
+import (
+	"fmt"
+	"sort"
+)
+
+// The default Priority is the last index. The last index is not required in the prioirty list, thus any elements added
+// to the struct will be added to the default priority (last).
+const defaultPrioirty = -1
+
+type PriorityValue interface {
+	GetPriority() string
+}
+
+// PrioritySortedStruct sorts elements by prioirity. Priority lists are unordered.
+// Needs a priority list and a get priority function.
+// If a priority does not exist for an added element, it is added to the lowest priority.
+type PrioritySortedStruct struct {
+	// prioritySets is the set lists for priorities, where P is the type used for priority.
+	prioritySets map[int][]string
+	// priorityMap maps the value to the priority.
+	priorityMap map[string]int
+	// elements are the map of elements structured by their priority.
+	elements []map[int]ResponseWatch
+	// numberOfPriorities is the number of priorities in the prioritySets
+	numberOfPriorities int
+	// nextUniqueElementIndex is the next element index to be inserted.
+	nextUniqueElementIndex int
+}
+
+// PriorityIndex is the priority and index used to locate items.
+type PriorityIndex struct {
+	Priority int
+	Index    int
+}
+
+func NewPrioritySortedStruct(prioritySets map[int][]string) *PrioritySortedStruct {
```

Comment:
Why are we keeping strings here when we could just keep a watch like a map[int][]watch or a [][]Watch

Reply:
The strings are the `Cluster` and `Listener`. Look at the Generics in this [PR](https://github.com/solo-io/go-utils/pull/483). The type should be comparable.

Reply:
The items that the collection contains are ResponseWatches.
This is the `elements` property.